### PR TITLE
Add support for KTX image format so that we can use Basis Universal for GLTF

### LINKED ISSUE #5873: Support KHR_texture_basisu extension when loading glTF scenes
### Describe the project you are working on

A metaverse which allows users to upload their own model, trade them, place them and construct world together, which means there will be hundreds and thousands model will be placed in one scene, so the file size of each model becomes crucial.

### Describe the problem or limitation you are having in your project

Currently godot cannot load glb file which is compressed with KHR_texture_basisu texture, which will increase model download/load time, and also will occupy much more storage space.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The KHR_texture_basisu extension of GLTF 2.0 could compress all the textures to KTX2, which is GPU compatible with very small size.
Invoke `gltfpack -tc -i input.glb -o output.glb` will compress the model from 180MB to 28.5MB with nearly no notable difference which is a huge improvement.  

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Compile GLTF model with KHR_texture_basisu enabled and support KTX 2.0 texture format.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, cannot worked around with script.

### Is there a reason why this should be core and not an add-on in the asset library?

GLTF support is i core itself.

### PR DESCRIPTION
**Pull request by acazuc**
Bugsquad Edit: This implements KTX support with Basisu gltf extension and fixes: https://github.com/godotengine/godot-proposals/issues/5873 The final binary size is ~100KB added.

Hello,

This merge request allows KTX file format texture loading
There are some limitations to supported KTX files: 2d images only, no cubemap, no array, no de-padding, only a limited list of godot Image formats
Decoding is done using [libktx](https://github.com/KhronosGroup/KTX-Software)
I tested ktx files from [ktxtext](https://github.com/sevmeyer/ktxtest.git), most useful formats are working (rgb8, rgba8, astc, bc1-7, and some others), invalid / unsupported files are handled with error messages
KTX uses 4-bytes padding, and no de-padding is done here:
- if padding is used on non-first mipmap, only the first mipmap is used
- if padding is used on the first mipmap, file isn't loaded

I guess it does help for https://github.com/godotengine/godot-proposals/issues/5873

Compilation & tests has been done in amd64 target

Regards,

### CODE REVIEW

**Comment by acazuc** [comment]
The biggest remaining file is `vulkan_core.h` (11k lines), which is also located in `thirdparty/vulkan`
Would it be ok for you to add `thirdparty/vulkan/include` as an include path to build `thirdparty/libktx` in `modules/ktx/SCsub` ?

**Comment by fire** [comment]
I will try to review. Trying get some more opinions from the godot engine team members

**Comment by fire** [comment]
```
Error: D:\a\godot\godot\thirdparty\libktx\lib\gl_format.h(95): error C2220: the following warning is treated as an error
Warning: D:\a\godot\godot\thirdparty\libktx\lib\gl_format.h(95): warning C4005: 'NOMINMAX': macro redefinition
```

Weird error. Might have to hotpatch and send upstream?

**Comment by RevoluPowered** [comment]
Hey, I was asked to review this by @fire and just wanted to add some supporting info:

**Slides from Khronos**
https://www.khronos.org/assets/uploads/apis/KTX-2.0-Launch-Overview-Apr21_.pdf

The goal here as far as I can understand is that we want to make exported games smaller and also we would like to reduce the amount of resource duplication.

Basically apple/android uses ETC2 or ASTC. 
Windows/Linux use BTPC, DXT5 formats.

This means you export your game it includes them on windows both of them, potentially making things huge.

So that means you import one texture: and it becomes 3 potentially. This unifies in a single format "basis universal" which can easily transcode into those formats. This moves texture import time outside of the editor time, without compromising performance/quality.

**For non subject matter experts:** smaller game export, faster editor import time, faster boot time because the dataset for the export is smaller.

File: thirdparty/libktx/Apache-2.0.txt [hunk 1]
```diff
@@ -0,0 +1,208 @@
+Apache License
```
Comment (RevoluPowered) [inline thirdparty/libktx/Apache-2.0.txt:1]: @akien-mga is apache 2 compatible with MIT? (for godot)

**Review by RevoluPowered — APPROVED** [review]
LGTM

The only portion I dislike is the massive switch case but it looks simple enough and is OK that it  is explicit, it's also fast so, I should not complain really 😅 

Maybe recommend writing a very simple unit test for this if you can.

**Comment by RevoluPowered** [comment]
* Needs some docs on how to use this

**Comment by Calinou** [comment]
Does this PR allow loading arbitrary KTX images from the filesystem or a buffer in an exported project, similar to `Image.load_png_from_buffer()`? As far as I know, this is required to work with the run-time glTF loader.

File: thirdparty/libktx/Apache-2.0.txt (hunk 1 above)
Comment (Calinou) [reply thirdparty/libktx/Apache-2.0.txt:1]: Yes, we already use Apache 2-licensed dependencies.

**Comment by acazuc** [comment]
It doesn't add arbitrary loading like png, it only adds support for KTX file the same way DDS files are handled (as imported texture files)
I guess KTX2 with basisu loading for glTF could be handled another way, maybe using ktx2_transcoder class from `thirdparty/basis_universal` the same way `modules/basis_universal` does handle basisu files

**Comment by fire** [comment]
Do you know how much effort it is to load_ktx_from_buffer and load_png_from_file? Was curious. We'd need to add the api.

**Comment by acazuc** [comment]
A commit was added for KHR_texture_basisu support in glTF module
I tested the import of https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/StainedGlassLamp/glTF-KTX-BasisU with success

NB: there is a shadow dependency in `modules/ktx/SCsub` for basisu transcoder: `thirdparty/basis_universal/transcoder/basisu_transcoder.cpp` is not added to source because it does collide with `modules/basis_universal/SCsub` target and I don't know how to solve it gracefully.

Sorry for the back and forthes of this review

**Comment by fire** [comment]
I'll try asking for support. Aside from that, there are some whitespace cicd errors.

**Comment by aaronfranke** [comment]
@acazuc With that test asset you posted, when I try to load `StainedGlassLamp_base_basecolor.ktx2` (not as part of a GLTF, just by double-clicking on the texture by itself) I get this error:

```
 Unsupported format 0 of KTX2 texture file 'res://StainedGlassLamp/glTF-KTX-BasisU/StainedGlassLamp_base_basecolor.ktx2'.
  Invalid image
  Failed loading resource: res://StainedGlassLamp/glTF-KTX-BasisU/StainedGlassLamp_base_basecolor.ktx2. Make sure resources have been imported by opening the project in the editor at least once.
  editor/editor_node.cpp:1263 - Condition "!res.is_valid()" is true. Returning: ERR_CANT_OPEN
```

**Comment by Calinou** [comment]
> It doesn't add arbitrary loading like png, it only adds support for KTX file the same way DDS files are handled (as imported texture files)

Note that https://github.com/godotengine/godot/pull/69085 implements run-time DDS loading.

**Comment by acazuc** [comment]
@aaronfranke that is strange, did you used the last commit (at least [6271fdd](https://github.com/godotengine/godot/pull/76572/commits/6271fdd6443daf067900c16fcbfea01d8c7eaffb)) ?
This error did happen before this commit because basisu transcode wasn't supported

**Comment by aaronfranke** [comment]
@acazuc I just re-tested, this time with 8ca3bbc0e77ddcfb35c64878889c574a19ff7675, and it works. Although interestingly, the PNG and JPG version does not work. It might be the same issue as #76691.

---

Well, now that I've tested that this works... As for giving this PR a review, I'm not really sold on this being in core due to the extremely large amount of code required. 30k lines is a lot.

Regardless of whether we want this to be in core or not, I want to modify the code in GLTFDocument to allow GLTFDocumentExtension classes to extend the importer with arbitrary image formats, instead of just adding KTX to the current hard-coded code. As a first test I would like to implement `EXT_texture_webp` via GLTFDocumentExtension since Godot already includes all of the code required to save and load WebP files, then afterwards we can copy-paste that code as a template for KTX (or possibly just port that code to GDExtension if we decide to not put KTX in core). I spent today writing the code to do this and it seems to load WebP files into memory fine. However, I am unable to fully test it because I can't even get PNG files to work on master, see #76691.

**Comment by fire** [comment]
Is it correct that thirdparty/libktx/lib/dfdutils/vulkan/vulkan_core.h is a duplicate because it costs 10,000 lines of code?

**Comment by fire** [comment]
@acazuc Do you require some help with removing the "GHA / 🐧 Linux / Editor w/ Mono (target=editor) (pull_request) Failing after 4m" Github action test failure?

**Comment by fire** [comment]
Can you update this pr based on https://github.com/godotengine/godot/pull/76895? I am not sure how long the merge window for 4.1 lasts, but it's arriving soon.

**Comment by fire** [comment]
@Calinou What's the process for checking binary size comparisons?

File: modules/gltf/extensions/gltf_document_extension_texture_ktx.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "gltf_document_extension_texture_ktx.h"
+
+#include "scene/3d/area_3d.h"
```
Comment (aaronfranke) [inline modules/gltf/extensions/gltf_document_extension_texture_ktx.cpp:33]:
```suggestion
```

Sorry about that, this unnecessary include was included by me in the WebP file too, copy-paste error. You can also delete this unnecessary include in that file, that would be appreciated.

**Review by aaronfranke — APPROVED** [review]
In terms of the code organization: This looks good to me, the code in the GLTF module is clean, the module can be disabled just with a scons option, and the Image code is similar to the code for the other image formats.

In terms of whether we want this in the engine or not: I don't know, 100 KiB of optional binary size is not that bad though. To me it depends on how often this will be used. I will leave the decision to others like @fire and @akien-mga. In the meantime I'll hit the approve button myself.

**Review by fire — APPROVED** [review]
https://github.com/KhronosGroup/glTF-Sample-Models/tree/4ca06672ce15d6a27bfb5cf14459bc52fd9044d1/2.0/FlightHelmet/glTF-KTX-BasisU

Tested with this basisu gltf file.

image

It's the scope of this pr to allow ktx2 import, but it's outside the scope of this pr to enforce things like only basisu for a gltf.

Great work!

Note:

In my view 100KB is ok to use for this, thoughts?

**Comment by fire** [comment]
Can you refresh the pr?

**Comment by Calinou** [comment]
Binary size comparison for a Linux x86_64 stripped release export template with LTO:

```
70,341,920  godot.linuxbsd.template_release.x86_64.master
70,448,800  godot.linuxbsd.template_release.x86_64.pr
```

**Review by reduz — APPROVED** [review]

**Comment by akien-mga** [comment]
Just noting for reviewers that I still need to check the thirdparty code integration, but otherwise this should be good to merge once I've done that.

**Comment by akien-mga** [comment]
Did a quick review pass, the integration of the thirdparty library is pretty good! Great job trimming down all the files we don't need to keep it minimal.

Here's a diff with some further changes to properly document the copyright, and handle the fact that Godot can be compiled without `basis_universal` or without Vulkan.

```diff
diff --git a/COPYRIGHT.txt b/COPYRIGHT.txt
index 582784d78e..904b2adca2 100644
--- a/COPYRIGHT.txt
+++ b/COPYRIGHT.txt
@@ -256,6 +256,12 @@ Comment: jpeg-compressor
 Copyright: 2012, Rich Geldreich
 License: public-domain or Apache-2.0
 
+Files: ./thirdparty/libktx/
+Comment: KTX
+Copyright: 2013-2020, Mark Callow
+  2010-2020 The Khronos Group, Inc.
+License: Apache-2.0
+
 Files: ./thirdparty/libogg/
 Comment: OggVorbis
 Copyright: 2002, Xiph.org Foundation
diff --git a/modules/ktx/SCsub b/modules/ktx/SCsub
index 9507d5c4c8..302a28f94d 100644
--- a/modules/ktx/SCsub
+++ b/modules/ktx/SCsub
@@ -11,7 +11,6 @@ thirdparty_obj = []
 
 thirdparty_dir = "#thirdparty/libktx/"
 thirdparty_sources = [
-    "lib/basis_transcode.cpp",
     "lib/checkheader.c",
     "lib/filestream.c",
     "lib/hashlist.c",
@@ -31,12 +30,19 @@ thirdparty_sources = [thirdparty_dir + file for file in thirdparty_sources]
 
 env_ktx.Prepend(CPPPATH=[thirdparty_dir + "include"])
 env_ktx.Prepend(CPPPATH=[thirdparty_dir + "utils"])
-env_ktx.Prepend(CPPPATH=[thirdparty_dir + "lib"])
-env_ktx.Prepend(CPPPATH=[thirdparty_dir + "lib/dfdutils"])
 env_ktx.Prepend(CPPPATH=[thirdparty_dir + "other_include"])
-env_ktx.Prepend(CPPPATH=["#thirdparty/basis_universal"])
-env_ktx.Prepend(CPPPATH=["#thirdparty/vulkan/include"])
-env_ktx.Prepend(CPPFLAGS=["-DKHRONOS_STATIC=1"])
+
+if env["module_basis_universal_enabled"]:
+    thirdparty_sources += [thirdparty_dir + "lib/basis_transcode.cpp"]
+    env_ktx.Prepend(CPPPATH=["#thirdparty/basis_universal"])
+
+if env["vulkan"]:
+    env_ktx.Prepend(CPPPATH=["#thirdparty/vulkan/include"])
+else:
+    # Falls back on bundled `vkformat_enum.h`.
+    env_ktx.Append(CPPDEFINES=["LIBKTX"])
+
+env_ktx.Append(CPPDEFINES=[("KHRONOS_STATIC", 1)])
 
 env_thirdparty = env_ktx.Clone()
 env_thirdparty.disable_warnings()
diff --git a/thirdparty/README.md b/thirdparty/README.md
index 611f63b02a..a4e4294a77 100644
--- a/thirdparty/README.md
+++ b/thirdparty/README.md
@@ -303,7 +303,7 @@ Files extracted from upstream source:
 
 - Upstream: https://github.com/KhronosGroup/KTX-Software
 - Version: 4.1.0 (d7255fe73cd53b856731ceb9f2c279181d0dbbca, 2023)
-- License: MIT
+- License: Apache-2.0
 
 Files extracted from upstream source:
 
@@ -315,7 +315,7 @@ Files extracted from upstream source:
 - `other_include/KHR/*`
 - ifndef-protect NOMINMAX define in `lib/gl_format.h`
 - remove `basisu/` prefix from `thirdparty/libktx/lib/basis_transcode.cpp` basisu includes
-- comment VK\_FORMAT\_ASTC_\*x\*x\*\_UNORM\_BLOCK\_EXT cases in lib/dfdutils/vk2dfd.inl
+- comment `VK_FORMAT_ASTC_*x*x*_UNORM_BLOCK_EXT` cases in `lib/dfdutils/vk2dfd.inl`
 
 
 ## libogg
```

The instructions in `thirdparty/README.md` are incomplete, as they suggest to copy over all files from the upstream repo, which is not what you ended up doing after trimming things down.

We need to have precise documentation on which files to copy or not, so that future updates can be done easily. (E.g. there's already version 4.2.1 upstream which we might want to update too eventually).

And this part might be best described as a `.patch` file so that it can be re-applied on updates:
```
- ifndef-protect NOMINMAX define in `lib/gl_format.h`
- remove `basisu/` prefix from `thirdparty/libktx/lib/basis_transcode.cpp` basisu includes
- comment `VK_FORMAT_ASTC_*x*x*_UNORM_BLOCK_EXT` cases in `lib/dfdutils/vk2dfd.inl
```

File: core/io/image.cpp [hunk 3]
```diff
@@ -3873,6 +3875,14 @@ Error Image::load_dds_from_buffer(const Vector<uint8_t> &p_array) {
 	return _load_from_buffer(p_array, _dds_mem_loader_func);
 }
 
+Error Image::load_ktx_from_buffer(const Vector<uint8_t> &p_array) {
+	ERR_FAIL_NULL_V_MSG(
+			_ktx_mem_loader_func,
+			ERR_UNAVAILABLE,
+			"The KTX module isn't enabled. Recompile the Godot editor or export template binary with the `module_bmp_enabled=yes` SCons option.");
```
Comment (akien-mga) [inline core/io/image.cpp:3882]:
```suggestion
			"The KTX module isn't enabled. Recompile the Godot editor or export template binary with the `module_ktx_enabled=yes` SCons option.");
```

File: modules/ktx/texture_loader_ktx.cpp [hunk 4]
```diff
@@ -0,0 +1,561 @@
...
+KTX_error_code ktx_read(ktxStream *stream, void *dst, const ktx_size_t count) {
+	Ref<FileAccess> *f = reinterpret_cast<Ref<FileAccess> *>(stream->data.custom_ptr.address);
+	(*f)->get_buffer(reinterpret_cast<uint8_t *>(dst), count);
+	return KTX_SUCCESS;
+}
+
+KTX_error_code ktx_skip(ktxStream *stream, const ktx_size_t count) {
+	Ref<FileAccess> *f = reinterpret_cast<Ref<FileAccess> *>(stream->data.custom_ptr.address);
+	for (ktx_size_t i = 0; i < count; ++i)
+		(*f)->get_8();
```
Comment (akien-mga) [inline modules/ktx/texture_loader_ktx.cpp:49]:
```suggestion
	for (ktx_size_t i = 0; i < count; ++i) {
		(*f)->get_8();
	}
```

File: modules/ktx/texture_loader_ktx.cpp [hunk 5]
```diff
@@ -0,0 +1,561 @@
...
+				case GL_COMPRESSED_RG_RGTC2:
+					format = Image::FORMAT_RGTC_RG;
+					break;
+				case GL_COMPRESSED_RGB_BPTC_UNSIGNED_FLOAT:
+					format = Image::FORMAT_BPTC_RGBFU;
+					break;
+				case GL_COMPRESSED_RGBA_BPTC_UNORM:
+					format = Image::FORMAT_BPTC_RGBA;
+					break;
+#if 0
```
Comment (akien-mga) [inline modules/ktx/texture_loader_ktx.cpp:210]: Needs a comment to explain why it's commented out.

File: modules/ktx/texture_loader_ktx.cpp [hunk 6]
```diff
@@ -0,0 +1,561 @@
...
+			ktxTexture2 *ktx_texture2 = reinterpret_cast<ktxTexture2 *>(ktx_texture);
+			if (ktx_texture2->vkFormat == VK_FORMAT_UNDEFINED) {
+				if (!ktxTexture2_NeedsTranscoding(ktx_texture2)) {
+					ktxTexture_Destroy(ktx_texture);
+					ERR_FAIL_V_MSG(Ref<Resource>(), "Invalid VK_FORMAT_UNDEFINED of KTX2 texture file '" + p_path + "'.");
+				}
+				ktx_transcode_fmt_e ktxfmt;
+				switch (ktxTexture2_GetNumComponents(ktx_texture2)) {
+					case 1: {
+						if (ktxTexture2_GetOETF_e(ktx_texture2) == KHR_DF_TRANSFER_SRGB) { /* XXX srgb native support */
```
Comment (akien-mga) [inline modules/ktx/texture_loader_ktx.cpp:263]: What does "XXX" mean?

File: modules/ktx/texture_loader_ktx.cpp [hunk 7]
```diff
@@ -0,0 +1,561 @@
...
+			ERR_FAIL_V_MSG(Ref<Resource>(), "Invalid KTX texture file '" + p_path + "'.");
+		}
+		int prev_size = src_data.size();
+		src_data.resize(prev_size + mip_size);
+		memcpy(src_data.ptrw() + prev_size, ktxTexture_GetData(ktx_texture) + offset, mip_size);
+	}
+
+	Ref<Image> img = memnew(Image(width, height, mipmaps - 1, format, src_data));
+	if (srgb)
+		img->srgb_to_linear();
```
Comment (akien-mga) [inline modules/ktx/texture_loader_ktx.cpp:501]: Please make sure to always use braces.
```suggestion
	if (srgb) {
		img->srgb_to_linear();
	}
```

File: modules/ktx/texture_loader_ktx.cpp [hunk 8]
```diff
@@ -0,0 +1,561 @@
...
+	Ref<ImageTexture> texture = ImageTexture::create_from_image(img);
+	return texture;
+}
+
+static Ref<Image> _ktx_mem_loader_func(const uint8_t *p_ktx, int p_size) {
+	Ref<FileAccessMemory> f;
+	f.instantiate();
+	f->open_custom(p_ktx, p_size);
+	Error err;
+	Ref<Image> img = load_from_file_access(f, &err, "memory");
```
Comment (akien-mga) [inline modules/ktx/texture_loader_ktx.cpp:538]: What is `"memory"`? Seems only used in error messages and that wouldn't make much sense to the user.

File: thirdparty/libktx/LICENSE.adoc [hunk 9]
```diff
@@ -0,0 +1,10 @@
+= LICENSE file for the KhronosGroup/dfdutils project
```
Comment (akien-mga) [inline thirdparty/libktx/LICENSE.adoc:1]: With two license files it's unclear which is which, this one should be renamed to `LICENSE.dfdutils.adoc`.

**Review by akien-mga — CHANGES_REQUESTED** [review]

File: modules/ktx/texture_loader_ktx.cpp [hunk 10]
```diff
@@ -0,0 +1,561 @@
...
+}
+
+KTX_error_code ktx_getsize(ktxStream *stream, ktx_size_t *const size) {
+	Ref<FileAccess> *f = reinterpret_cast<Ref<FileAccess> *>(stream->data.custom_ptr.address);
+	*size = (*f)->get_length();
+	return KTX_SUCCESS;
+}
+
+void ktx_destruct(ktxStream *stream) {
+	(void)stream;
```
[the remark this answers is not in the document]
Comment (acazuc) [reply modules/ktx/texture_loader_ktx.cpp:78]: it doesn't do anything (the (void) is here to silent compiler warning, and this function seems to be required for the destruct pointer in ktxStream)

File: modules/ktx/texture_loader_ktx.cpp [hunk 11]
```diff
@@ -0,0 +1,561 @@
...
+			ktxTexture_Destroy(ktx_texture);
+			ERR_FAIL_V_MSG(Ref<Resource>(), "Unsupported version KTX texture file '" + p_path + "'.");
+			break;
+	}
+
+	Vector<uint8_t> src_data;
+
+	/* KTX use 4-bytes padding, don't use mipmaps if padding is effective
+	 * XXX unpad dynamically
+	 */
```
[the remark this answers is not in the document]
Comment (acazuc) [reply modules/ktx/texture_loader_ktx.cpp:460]: Yes it is

[the remark this answers is not in the document]
Comment (akien-mga) [reply modules/ktx/texture_loader_ktx.cpp:460]: I would suggest using `TODO` then, we don't use `XXX` for this anywhere in the codebase.

File: modules/ktx/texture_loader_ktx.cpp (hunk 8 above)
Comment (akien-mga) [reply modules/ktx/texture_loader_ktx.cpp]: I mean the `p_path` parameter in this function is unused, so it should be removed.

**Review by akien-mga — APPROVED** [review]
Looks really good now!

The only bit I'm missing from my earlier review is improving the update instructions in `thirdparty/README.md`, to know which files to copy and which ones to leave out.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Amazing work, thanks! Congrats for your first merged Godot contribution :tada:

**Comment by nikitalita** [comment]
@acazuc What do you mean by the `TODO: ETC compression is bogus` comments? I ask because I've attempted to load ETC-compressed ktx images by turning on those pragmas, and while they do load and display correctly in the editor, the thumbnail is garbage pixels. Additionally, when re-saving it, it saves it as a texture with similarly garbage pixels.
image
image

**Comment by clayjohn** [comment]
@nikitalita Thumbnails being broken is a sign that decompression is broken. You could test that theory by manually loading the texture in GDScript, decompressing, then displaying the resulting image

**Comment by acazuc** [comment]
@nikitalita I had the same problem with ETC compressed images in the KTX test suite
I didn't know how to fix that and ETC is rarely used, so I added that ifdef to mark ETC as "broken, but may be solved with some investigations"

**Comment by leanmendoza** [comment]
@acazuc @nikitalita Have you noticed darker textures for those that `srgb` turn **true**?

After commenting this convertion, it works fine (at least for the texture I see darker)
image
https://github.com/godotengine/godot/blob/0c45ace151f25de2ca54fe7a46b6f077be32ba6f/modules/ktx/texture_loader_ktx.cpp#L500

Original texture 
image


Darker texture
image


texture zipped
[QEbvwX_0.zip](https://github.com/user-attachments/files/17879545/QEbvwX_0.zip)

**Comment by fire** [comment]
@leanmendoza, can you make an issue for this? This will get lost here.

**Comment by leanmendoza** [comment]
> @leanmendoza, can you make an issue for this? This will get lost here.

Yes, I first wanted to give a try here, but done https://github.com/godotengine/godot/issues/99589

**Pull request merged**